Added support for DB-API 2.0 compatibility, missing set_trace_callback proxy

shipmints opened the pull request:
### Description

First, thank you to John and the other contributors for this library. It is tight and well written and easy to use.

1. Added support to Cursor to return self to ensure compatibility with DB-API 2.0 reliant code. This style uses cursors as the main point of execution control vs. the connection. Retrofitting aiosqlite into that style of code wasn't possible without this support.

2. Added set_trace_callback proxy support which was missing.

Note: We did not add additional tests for Cursor support. The existing tests are unaffected. This code is deployed in our production environment already. An additional test for set_trace_callback was also not added. Apologies for running out of time for these but we do know the code works.

Fixes: #60, #61

jreese commented:
Thanks! Can you resolve the type checking errors?

shipmints commented:
I corrected the method return type hints. Thanks for pointing that out.

shipmints commented:
I added a correction for the (antiquated) lint line-length check.

I do not understand the other complaint:

aiosqlite/core.py:113:0: R0904: Too many public methods (21/20) (too-many-public-methods)

shipmints commented:
I hacked in one extra allowable public method to .pylint. Seems a bit arbitrary to limit to 20 or even 21. Perhaps a power of two would be more fun.

shipmints commented:
I have to leave the remaining lint/black/formatting error failures in your hands.

jreese commented:
Hi, can you rebase onto latest master, and follow the [contributers guide](https://github.com/jreese/aiosqlite/blob/master/CONTRIBUTING.md) to make sure source files are correctly formatted? I ensured formatting/tests/lint are all passing on latest master.

shipmints commented:
Working on this now. FYI, "make format" changed a bunch of files and the changes indicate that perhaps this isn't desirable or perhaps it just hasn't been run in a long time. Examples below. Please advise. If this is what you want, I'll rebase them. If not, let me know what you'd like.

```
$ git status
...
	modified:   aiosqlite/__init__.py
	modified:   aiosqlite/core.py
	modified:   aiosqlite/tests/perf.py
	modified:   aiosqlite/tests/smoke.py
```

```
git diff aiosqlite/__init__.py 
diff --git a/aiosqlite/__init__.py b/aiosqlite/__init__.py
index a612117..9925884 100644
--- a/aiosqlite/__init__.py
+++ b/aiosqlite/__init__.py
@@ -4,21 +4,21 @@
 """asyncio bridge to sqlite3"""
 
 from sqlite3 import (  # pylint: disable=redefined-builtin
+    DatabaseError,
+    Error,
+    IntegrityError,
+    NotSupportedError,
+    OperationalError,
+    ProgrammingError,
+    Row,
+    Warning,
     register_adapter,
     register_converter,
     sqlite_version,
     sqlite_version_info,
-    Row,
-    Warning,
-    Error,
-    DatabaseError,
-    IntegrityError,
-    ProgrammingError,
-    OperationalError,
-    NotSupportedError,
 )
 
-from .core import connect, Connection, Cursor
+from .core import Connection, Cursor, connect
 
 __version__ = "0.12.0"
 __all__ = [
```

```
$ git diff aiosqlite/core.py 
diff --git a/aiosqlite/core.py b/aiosqlite/core.py
index 91e9212..eefe66a 100644
--- a/aiosqlite/core.py
+++ b/aiosqlite/core.py
@@ -8,10 +8,9 @@ Core implementation of aiosqlite proxies
 import asyncio
 import logging
 import sqlite3
-
 from functools import partial
 from pathlib import Path
-from queue import Queue, Empty
+from queue import Empty, Queue
 from threading import Thread
 from typing import Any, Callable, Generator, Iterable, Optional, Tuple, Type, Union
 
@@ -50,8 +49,9 @@ class Cursor:
         await self._execute(self._cursor.execute, sql, parameters)
         return self
 
-    async def executemany(self, sql: str,
-                          parameters: Iterable[Iterable[Any]]) -> "Cursor":
+    async def executemany(
+        self, sql: str, parameters: Iterable[Iterable[Any]]
+    ) -> "Cursor":
         """Execute the given multiquery."""
         await self._execute(self._cursor.executemany, sql, parameters)
         return self
@@ -299,9 +299,7 @@ class Connection(Thread):
     ) -> None:
         await self._execute(self._conn.set_progress_handler, handler, n)
 
-    async def set_trace_callback(
-        self, handler: Callable
-    ) -> None:
+    async def set_trace_callback(self, handler: Callable) -> None:
         await self._execute(self._conn.set_trace_callback, handler)
```

jreese commented:
Those changes were already resolved/committed on the master branch, so rebasing should remove the need for them in your branch.

jreese commented:
More specifically, the import sorting changes should already be in the master branch. The changes to lines you touched are desired.

shipmints commented:
I think this is all set now. Please let me know of any issues. Thanks, again, for a nice library.

shipmints commented:
Hi. Let me know if there's anything else I need to do with this.

jreese commented:
I was trying to add a test case for this, and started to realize that I'm not sure what your expected use cases is.  I suspect it won't work the way you want it to right now, assuming you want to do something like:

```python
async with db.cursor() as cursor:
    await cursor.execute(...).execute(...)
```

This won't work because you have to await each call to `execute()` before you would get the cursor back, meaning to chain calls correctly you would need to instead do:

```python
async with db.cursor() as cursor:
    await (await db.execute(...)).execute(...)
```

This is not nearly as readable as the first example, and I don't think it provides anything useful for most users. 

The `set_trace_callback()` part looks fine otherwise, though I would prefer to see a test case for that, and would be ready to merge if you split that into a new pull request.

shipmints commented:
The cursor-based primacy comes straight from the Python standard API which should be followed. An obvious use case would be chaining transactions under a single cursor which can't be done using the connection-oriented API e.g.,

```
# pseudocode
with cur:
  cur.execute('savepoint s1')
  some_other_nested_thing(cur)
  cur.execute('release savepoint s1')
```

We've moved on to other projects. It would have been good for you to provide this feedback a week ago. If you really, really need us to split the one-liner trace code from the three-lines changed to handle cursor primacy, we can do that. Seems like "busy work," though.

jreese commented:
I'm not opposed to these changes since they replicate the implementation details of the `sqlite3` module, but I would like to have reasonable test cases that can cover the appropriate use cases and guarantee that the functionality is working as expected. To that extent, I'm still not sure how returning the cursor from its own methods will help you when you need to await the methods before you would get the cursor value back. 

[PEP 249 (DB API 2.0)](https://www.python.org/dev/peps/pep-0249/#id15) specifically says "Return values are not defined" for the `.execute*()` methods. Your example code also does not depend on the return value of these methods, so I'm not sure what a valid test case would look like, other than expecting to get the cursor back after awaiting the methods. Like I mentioned in my previous comment, I'm not convinced that method chaining is a viable use case for async methods. But if you believe otherwise, I would appreciate a more detailed example that depends on the return value of cursors.

Do keep in mind that while this is a free library that I built and support in my spare time, I do want to maintain a high level of quality, ensure that CI jobs are passing, and so on. I don't always have the time or ability to respond immediately, and I rarely have time to expend effort on finishing others' pull requests for them. If you don't have a vested interest in this functionality, then you are likewise under no obligation to continue working on it.

shipmints commented:
I hear you, I'm with you. Willing to help keep aiosqlite great.

The existing aiosqlite implementation of Connection.execute returns a Cursor which is very useful. Let me get back into the code and see if, perhaps, we've used/abused the sqlite API ourselves and have been relying on behaviors that are exposed even by aiosqlite. 

According to https://docs.python.org/3.8/library/sqlite3.html

"execute(sql[, parameters])...This is a nonstandard shortcut..." yet is supported in aiosqlite and returns a cursor per:

```
    async def execute(self, sql: str, parameters: Iterable[Any] = None) -> Cursor:
...
        return Cursor(self, cursor)
```

Just saying. This is the behavior we'd like to see consistent in Cursor.execute.

shipmints commented:
I've taken a closer look at the team's usage, and they do depend on the sqlite default behavior of execute returning its own cursor. It was this behavior they were looking to have replicated (even if Python's own documentation is inconsistent) in aiosqlite. I can have them change the behavior of the code. It's not in many places. But let's first consider whether aiosqlite should replicate the non-wrapped API faithfully, even if inconsistently documented so that it's closer to a drop-in. I leave this in your hands. We can accept the change or not. I appreciate your attention to this and the effort to delve in.

shipmints commented:
I had the team undo their changes just in case you didn't want to adopt execute returning cursor.

For your reference, you can see the standard, if inconsistently documented, implementation here:

https://github.com/python/cpython/blob/master/Modules/_sqlite/cursor.c#L611

jreese commented:
I do think it makes sense to replicate the behavior of the underlying library, but let's make sure we have a basic test case covering this behavior, similar to how we validate the context manager functionality at https://github.com/omnilib/aiosqlite/blob/master/aiosqlite/tests/smoke.py#L135

Would that test literally just be asserting that the return value of `(await cursor.execute()) == cursor`, or is there something more nuanced to test? Is there an example we can put in the readme of how this enables better or cleaner usage of the library that wasn't possible before?

shipmints commented:
I think that simple test is all that's needed. Thanks, again, John.

jreese commented:
Manually merged.
